### PR TITLE
remove rnn infershape

### PR DESCRIPTION
**Pull request by jacquesqiao**
fix: https://github.com/PaddlePaddle/Paddle/issues/4555

### CODE REVIEW

File: paddle/operators/recurrent_op.cc [hunk 1]
```diff
@@ -30,21 +30,23 @@ using LoDTensor = framework::LoDTensor;
...
+  PADDLE_ENFORCE_NOT_NULL(input0);
+  seq_len_ = input0->GetMutable<LoDTensor>()->dims()[0];
+  PADDLE_ENFORCE_GT(seq_len_, 0);
+
+  CreateScopes(scope);
+  auto& step_scopes = GetStepScopes(scope);
+  rnn::SegmentInputs(step_scopes, arg_->inlinks, seq_len_);
+  InitMemories(step_scopes[0]);
+
+  for (size_t i = 0; i < seq_len_; i++) {
```
Comment (Superjom) [inline paddle/operators/recurrent_op.cc]: i -> stepid

File: paddle/operators/rnn/recurrent_op_utils.cc [hunk 2]
```diff
@@ -41,51 +41,45 @@ void SegmentInputs(const std::vector<Scope*>& step_scopes,
...
-      output->mutable_data<float>(platform::CPUPlace());
-      for (size_t j = 0; j < seq_len; j++) {
-        LoDTensor* step_output =
-            step_scopes[j]->FindVar(outlinks[i])->GetMutable<LoDTensor>();
-        // TODO(luotao02) data type and platform::DeviceContext() should set
-        // correctly
-        (output->Slice<float>(j, j + 1))
-            .CopyFrom<float>(*step_output, platform::CPUPlace());
-      }
+    auto step_scope_var = step_scopes[0]->FindVar(outlinks[i]);
```
Comment (Superjom) [inline paddle/operators/rnn/recurrent_op_utils.cc]: auto*

File: paddle/operators/recurrent_op.cc [hunk 3]
```diff
@@ -146,23 +144,22 @@ class RecurrentAlgorithmProtoAndCheckerMaker
 
 void RecurrentGradientAlgorithm::Run(
     const Scope& scope, const platform::DeviceContext& dev_ctx) const {
+  seq_len_ =
+      scope.FindVar(arg_->inlinks[0])->GetMutable<LoDTensor>()->dims()[0];
```
Comment (Superjom) [inline paddle/operators/recurrent_op.cc]: enforce not null scope.FindVar

File: paddle/operators/recurrent_op.cc (hunk 1 above)
Comment (jacquesqiao) [reply paddle/operators/recurrent_op.cc]: done

File: paddle/operators/recurrent_op.cc (hunk 3 above)
Comment (jacquesqiao) [reply paddle/operators/recurrent_op.cc]: done

File: paddle/operators/rnn/recurrent_op_utils.cc (hunk 2 above)
Comment (jacquesqiao) [reply paddle/operators/rnn/recurrent_op_utils.cc]: done

**Pull request merged**